chpl/chpl-service/src/main/java/gov/healthit/chpl/dao/impl/CriterionProductStatisticsDAOImpl.java

```diff
@@ -0,0 +1,117 @@
+    private static final long MODIFIED_USER_ID = -3L;
+
+    @Override
+    public List<CriterionProductStatisticsDTO> findAll() {
+        List<CriterionProductStatisticsEntity> result = this.findAllEntities();
+        List<CriterionProductStatisticsDTO> dtos = new ArrayList<CriterionProductStatisticsDTO>(result.size());
+        for (CriterionProductStatisticsEntity entity : result) {
+            dtos.add(new CriterionProductStatisticsDTO(entity));
+        }
+        return dtos;
+    }
+
+    @Override
+    public void delete(final Long id) throws EntityRetrievalException {
+        CriterionProductStatisticsEntity toDelete = getEntityById(id);
+
+        if (toDelete != null) {
+            toDelete.setDeleted(true);
+            toDelete.setLastModifiedUser(getUserId());
+            entityManager.merge(toDelete);
+        }
+
+    }
+
+    @Override
+    public CriterionProductStatisticsEntity create(final CriterionProductStatisticsDTO dto)
+            throws EntityCreationException, EntityRetrievalException {
+        CriterionProductStatisticsEntity entity = new CriterionProductStatisticsEntity();
+        entity.setProductCount(dto.getProductCount());
+        entity.setCertificationCriterionId(dto.getCertificationCriterionId());
+
+        if (dto.getDeleted() != null) {
+            entity.setDeleted(dto.getDeleted());
+        } else {
+            entity.setDeleted(false);
+        }
+
+        if (dto.getLastModifiedUser() != null) {
+            entity.setLastModifiedUser(dto.getLastModifiedUser());
+        } else {
+            entity.setLastModifiedUser(getUserId());
+        }
+        if (dto.getLastModifiedDate() != null) {
+            entity.setLastModifiedDate(dto.getLastModifiedDate());
+        } else {
+            entity.setLastModifiedDate(new Date());
+        }
+        if (dto.getCreationDate() != null) {
+            entity.setCreationDate(dto.getCreationDate());
+        } else {
+            entity.setCreationDate(new Date());
+        }
+
+        entityManager.persist(entity);
+        entityManager.flush();
+        return entity;
+    }
+
+    private List<CriterionProductStatisticsEntity> findAllEntities() {
+        Query query = entityManager
+                .createQuery("SELECT a from CriterionProductStatisticsEntity a where (NOT a.deleted = true)");
+        return query.getResultList();
+    }
+
+    private CriterionProductStatisticsEntity getEntityById(final Long id) throws EntityRetrievalException {
+        CriterionProductStatisticsEntity entity = null;
+
+        Query query = entityManager.createQuery(
+                "from CriterionProductStatisticsEntity a where (NOT deleted = true) AND (id = :entityid) ",
+                CriterionProductStatisticsEntity.class);
+        query.setParameter("entityid", id);
+        List<CriterionProductStatisticsEntity> result = query.getResultList();
+
+        if (result.size() > 1) {
```

Comment:
I think these bits of code that check if result.size() > 1 are in almost all the DAOs and they are from copying and pasting from Watson's days but I always thought they were kind of odd. How can there be more than one record with the same primary key? Not saying to remove because maybe it is a valid check, I just always thought they were weird!

Reply:
Fair. I can see some use for this kind of check in entities based on views, as we do sometimes have more than one entity in a view, but certainly seems odd for a query on a primary key. What do you think it ought to do? Just assume the "first" one is the right one? Or do we leave it as is?

Reply:
I guess I think the first one should also be the only one. Probably just leave it as it is though since it's a bit of code that's in most other places and not hurting anything. I just wanted to mention it!

Reply:
Fair enough.